components/inputs/input-color.js

```diff
@@ -0,0 +1,415 @@
+
+	}
+
+	async updated(changedProperties) {
+
+		super.updated(changedProperties);
+
+		if (changedProperties.has('label') || changedProperties.has('type')) this._validateLabel();
+
+	}
+
+	_getLabel() {
+		if (this.label !== undefined) return this.label;
+		if (this.type === 'background') return this.localize('components.input-color.backgroundColor');
+		if (this.type === 'foreground') return this.localize('components.input-color.foregroundColor');
+		return '';
+	}
+
+	_getOpener() {
+		if (this.readonly) {
+			return html`<div class="readonly-wrapper" id="opener" role="img" tabindex="0">${this._getSwatch()}</div>`;
+		}
+		const buttonClass = {
+			'd2l-button': true,
+			'd2l-dropdown-opener': this.launchType === 'dropdown',
+			'opened': this._opened
+		};
+		const ariaLabel = this._opened ? this._getTooltipLabel() : undefined;
+		const button = html`
+			<button id="opener" class="${classMap(buttonClass)}" aria-disabled="${ifDefined(this.disabled ? 'true' : undefined)}" aria-label="${ifDefined(ariaLabel)}" @click="${this._handleOpenDialog}">
+				${this._getSwatch()}
+				<svg xmlns="http://www.w3.org/2000/svg" width="10" height="6" fill="none" viewBox="0 0 10 6">
+					<path fill="#202122" d="M4.792 5.528a.733.733 0 0 1-.537-.223L.224 1.282a.745.745 0 0 1 0-1.065.751.751 0 0 1 1.057 0l3.51 3.511L8.303.218A.751.751 0 0 1 9.36 1.281L5.337 5.305a.753.753 0 0 1-.535.223h-.01Z"/>
+				</svg>
+			</button>`;
+		if (this.launchType === 'dialog') {
+			return button;
+		}
+		return html`
+			<d2l-dropdown @d2l-dropdown-close="${this._handleClose}" @d2l-dropdown-open="${this._handleOpenDropdown}">
+				${button}
+				<d2l-dropdown-content max-width="350" mobile-tray="left" no-padding ?opened="${this._opened}">
+					<slot @d2l-input-color-dropdown-close="${this._handleCloseRequest}"><div style="padding: 10px">Currently, color inputs only support a "launch-type" of "dropdown" from within the MVC framework. Use "launch-type" of "dialog" instead.</div></slot>
+				</d2l-dropdown-content>
+			</d2l-dropdown>`;
+	}
+
+	_getSwatch() {
+		const styles = {
+			backgroundColor: this.value ?? 'transparent'
+		};
+		const swatchClass = {
+			'swatch': true,
+			'swatch-transparent': (this.value === undefined)
+		};
+		const swatch = html`<div class="${classMap(swatchClass)}" style="${styleMap(styles)}"></div>`;
+		if (this.type === 'background' || this.type === 'foreground') {
+			const icon = (this.type === 'background') ? ICON_BACKGROUND : ICON_FOREGROUND;
+			return html`<div class="icon-wrapper">${icon}${swatch}</div>`;
+		}
+		return swatch;
+	}
+
+	_getTooltipLabel() {
+		const valueLabel = this.value !== undefined ? this.value : this.localize('components.input-color.none');
+		return `${this._getLabel()}: ${valueLabel}.`;
+	}
+
+	_handleClose() {
+		this._opened = false;
+		this.dispatchEvent(new CustomEvent(
+			'd2l-input-color-close', { bubbles: false, composed: false }
+		));
+	}
+
+	_handleCloseRequest(e) {
```

Comment:
It looks like this handler only runs if the dropdown is in play. Same for `_handleClose`. When the dialog is in play, when does `this._opened` get set to `false`?

Reply:
Good catch. `_opened` was really just for the dropdown and to hide the tooltip when it was open -- not needed for the dialog case since it's technically never "open".